### DISCUSSION TITLE
Helm releases are processed in batches of 4

### DISCUSSION THREAD

**Discussion by ghost** [fluxcd/flux2#3713] [Q&A] [answered]
I have noticed some performance impact when I switched from Helm Operator to Helm Controller, one of the reasons being that Helm Controller processes helm releases in batches of 4. i.e. if I had applied 10 helm releases to my cluster only 4 are processed in parallel and next 4 only start reconciliation once previous 4 is fully finished. (it's usually 20-30 seconds between batches)
Is this behavior expected ? and Is there a config I could change to increase batch size ?

**Comment by somtochiama** [comment, marked as answer]
The controllers have a `--concurrency` flags that defaults to 4, you can increase this to have the controller process more custom resources in parallel

**Reply by makkes** [reply]
Here's the full documentation: https://fluxcd.io/flux/cheatsheets/bootstrap/#increase-the-number-of-workers

**Reply by ghost** [reply]
thanks for a quick response. I will experiment with this setting.

**Reply by ghost** [reply]
Is there way I can verify the value of concurrency from inside of Helm Controller pod ?
```
containers:
  - args:
    - --events-addr=http://notification-controller.flux-system.svc.cluster.local./
    - --watch-all-namespaces
    - --log-level=info
    - --log-encoding=json
    - --enable-leader-election
    - --concurrent=20
 ```
  
This is pod settings, but helm releases are still reconciled in batches of 4.

**Reply by hiddeco** [reply]
It really is configured:

- https://github.com/fluxcd/helm-controller/blob/88a3beabd896be448e9c966b1f9aace6ec02c6c5/main.go#L206
- https://github.com/fluxcd/helm-controller/blob/88a3beabd896be448e9c966b1f9aace6ec02c6c5/controllers/helmrelease_controller.go#L131

**Reply by ghost** [reply]
Thanks again for quick replies. 
Configuration is indeed present, but I still noticed some strange behavior. Concurrency is only increased when:
* **new helm releases** are created (in this case they are processed according to new concurrency value)
* helm releases that had **READY status True** are updated (in this case updated releases are processed according to new concurrency value)

However if helm releases that had **READY status False** (or were previously failed for some reason) are updated  such releases are still reconciled with limit 4.
Is there a separate behavior or separate flag for failed release reconciliation parallelism ?
